Split Roblox-specific functionality into dedicated interface

voidedWarranties opened the pull request:
Closes #500 

## Outstanding issues

- [ ] Test Roblox-specific features, especially the studio plugin

voidedWarranties commented on src/AnalyzeCli.cpp:
```diff
@@ -329,10 +331,12 @@ int startAnalyze(const argparse::ArgumentParser& program)
         }
+
+        platform->handleRegisterDefinitions(frontend.globals, types::parseDefinitionsFileMetadata(*definitionsContents));
     }
 
-    types::registerInstanceTypes(frontend, frontend.globals, frontend.globals.globalTypes, fileResolver,
-        !program.is_used("--no-strict-dm-types") && client.configuration.diagnostics.strictDatamodelTypes);
+    platform->handleSourcemapUpdate(
+        frontend, frontend.globals, fileResolver, !program.is_used("--no-strict-dm-types") && client.configuration.diagnostics.strictDatamodelTypes);
```
Is working with `RobloxPlatform::instanceTypes` rather than `globals.globalTypes` okay? To be honest, I don't know the purpose of `TypeArena` enough to know what this changes.

JohnnyMorganz commented:
Sorry for the delay, haven't had a chance to review this yet. Will try to over the next few days.

JohnnyMorganz commented:
I went ahead and merged and resolved the conflicts between latest main and your branch. Will review in the next day or so.

JohnnyMorganz commented on src/operations/Completion.cpp:
```diff
@@ -474,8 +360,11 @@ static bool deprecated(const Luau::AutocompleteEntry& entry, std::optional<lsp::
 }
 
-static std::optional<lsp::CompletionItemKind> entryKind(const Luau::AutocompleteEntry& entry)
+std::optional<lsp::CompletionItemKind> WorkspaceFolder::entryKind(const Luau::AutocompleteEntry& entry)
```
minor q: could we leave this as static and instead just pass a platform ptr to the function? I try not to keep increasing the surface area of WorkspaceFolder

JohnnyMorganz commented on src/operations/Completion.cpp:
```diff
@@ -1004,15 +781,15 @@ std::vector<lsp::CompletionItem> WorkspaceFolder::completion(const lsp::Completi
                 if (auto typeReference = node->as<Luau::AstTypeReference>())
                     if (!typeReference->prefix)
-                        suggestImports(moduleName, position, config, *textDocument, items, /* includeServices: */ false);
+                        suggestImports(moduleName, position, config, *textDocument, items, /* isType: */ true);
```
nit: could we have a more descriptive name for `isType`? maybe `completingTypeReferencePrefix`, or something else

JohnnyMorganz commented on src/operations/Completion.cpp:
```diff
@@ -426,28 +330,10 @@ void WorkspaceFolder::suggestImports(const Luau::ModuleName& moduleName, const L
-                    textEdits.emplace_back(createServiceTextEdit(service, lineNumber, appendNewline));
-                }
-            }
+            platform->handleRequire(requirePath, lineNumber, isRelative, config, importsVisitor.get(), hotCommentsLineNumber, textEdits);
```
nit: `handleRequire` -> `handleRequireAutoImport`?

handleRequire could mean something else, such as actually mapping a `require(...)` to a path in a file resolver.

JohnnyMorganz reviewed (commented):
Overall, I think this is pretty awesome. The boundaries seem quite clear to me and extendable. Thank you for taking the time.

A general comment: I'm curious on the use cases of having platform specific Go To Definition, Hover, Inlay Hints and Signature Help. These seem very generalized and only need Luau, rather than platform specific handling. The only reason I ask is because the handlers you added are called very early on in the request, meaning the platform specific code loses out on all the generic processing we have already implemented. If platform-specific code is copy-pasting from the generic code, then I think the boundary is not at the right place. Do you have an example in mind here?

voidedWarranties commented:
> A general comment: I'm curious on the use cases of having platform specific Go To Definition, Hover, Inlay Hints and Signature Help.

My logic behind these was mostly adding special "syntax" within comments, like the `--- @` annotations I use for godot-luau-script. For greater control it may be better to move the handlers to the end rather than completely overriding default functionality. I'll try doing this when addressing the other reviews.

voidedWarranties commented on src/operations/ColorProvider.cpp:
```diff
@@ -242,7 +162,7 @@ lsp::DocumentColorResult WorkspaceFolder::documentColor(const lsp::DocumentColor
         return {};
 
-    DocumentColorVisitor visitor{textDocument};
+    DocumentColorVisitor visitor{textDocument, platform.get()};
     visitor.visit(sourceModule->root);
     return visitor.colors;
```
I think this makes sense. It's possible a platform may want to action on a string literal like "#ffffff" for example.

voidedWarranties commented on src/Workspace.cpp:
```diff
@@ -195,15 +198,13 @@ bool WorkspaceFolder::updateSourceMap()
         // We pass the same setting even when we are registering autocomplete globals since
         // the setting impacts what happens to diagnostics (as both calls overwrite frontend.prepareModuleScope)
-        types::registerInstanceTypes(frontend, frontend.globals, instanceTypes, fileResolver,
+        platform->handleSourcemapUpdate(frontend, frontend.globals, fileResolver,
```
My issue with abstracting away sourcemaps was that they seem to be quite tightly integrated with the main LSP, especially in `WorkspaceFileResolver`. Should the concept of `SourceNode`, virtual paths, etc. also be separated from the main LSP? Alternatively, is it possible for the concept of a sourcemap to be used by other platforms?

JohnnyMorganz replied:
Ah yeah, that is problematic. WorkspaceFileResolver built up a lot of cruft.

I wonder if we can generalise the idea of virtual and file paths, where "virtual" is treated as "LSP internal" in a sense. Platform should then only have two public API: a way to convert from file path to virtual, and then from virtual to real. The concept of "SourceNode" would then be Roblox-specific.

Or maybe it would even just be simpler for Platforms to implement their own file resolvers as a whole. FileResolvers are quite platform-specific.

I'm not sure how this would look without trying it out though. We could defer it to a future commit to prevent too much changing in one PR.

JohnnyMorganz commented:
I see (or maybe I'll see even better after looking at your godot implementation 😄)

What I would say is maybe we should start with a minimal interface that we extend on later, once we better understand how it all fits together. If you don't need to customise all of those operations yet, then maybe let's leave it out for now. But if you do, feel free to keep it in!

voidedWarranties commented on src/operations/ColorProvider.cpp:
```diff
@@ -255,37 +175,11 @@ lsp::DocumentColorResult LanguageServer::documentColor(const lsp::DocumentColorP
-    presentations.emplace_back(lsp::ColorPresentation{"Color3.fromHex(\"" + rgbToHex(rgb) + "\")"});
-
-    return presentations;
+    return platform->colorPresentation(params);
```
`platform` should always exist after config is received—is there any reason this would be called before then?

JohnnyMorganz replied:
Unsure. The only reason I asked is because you had a `if (!platform)` check for documentColor, so was wondering if we needed one for colorPresentation as well for completeness

voidedWarranties replied:
I see. I believe that check was a logical error.

voidedWarranties commented on tests/LuauExt.test.cpp:
```diff
@@ -1,6 +1,7 @@
 #include "doctest.h"
```
It is testing the base implementation of `FindImportsVisitor` which is in `LuauExt`. Should the suite be split then?

JohnnyMorganz replied:
Let's leave it for now, can reorganise later

voidedWarranties commented:
I've tried to address the review items as best I can—I've removed the `LSPPlatform` methods for the functionality I don't currently forsee handling, and I've also split sourcemaps from the main LSP as I believe it's in scope for this PR (unfortunately that seems to have doubled the size of the diff). As the dependency on `LSPPlatform` has grown more complex, I've added a system which should guarantee the LSP doesn't attempt to respond to requests requiring the platform before configuration has been received.

As before, I haven't had a chance to test Roblox functionality in-depth as I don't have an ideal setup to do so (I don't run Windows). If you could help testing the studio plugin, sourcemap, and any other relevant functionality that would be great.

JohnnyMorganz reviewed (commented):
> unfortunately that seems to have doubled the size of the diff

Yeah, unfortunately this is quite difficult to review. I trust that the code is mostly just copy-paste movements, rather than actual logic changes.

> As the dependency on LSPPlatform has grown more complex, I've added a system which should guarantee the LSP doesn't attempt to respond to requests requiring the platform before configuration has been received.

The reasoning behind this makes sense, but the postponed messages system seems a bit off for me. I already dislike how we postpone diagnostics until configuration, and now to do this for (almost) all LSP messages isn't great. Unfortuantely we do need a solution for it though.

I wonder if we should just abuse `initializationOptions` for this, to receive (at least the platform) config on startup. It is not technically correct, but I think it is important so that we don't have this complex logic here (and we might even be able to clean up the isConfigured check for diagnostics, but lets leave that for later). Of course, this requires work for clients, and we will need to handle default when this is not provided. (it looks like rust-analyzer does this already. relevant: https://github.com/microsoft/language-server-protocol/issues/1006, https://github.com/microsoft/language-server-protocol/issues/567)

I will also try to test out the code myself soon.

voidedWarranties commented:
> I wonder if we should just abuse `initializationOptions` for this

I did consider this; I believe the reason I didn't is that it seems impossible to use this method for more than one workspace with different platforms. Do you have any suggestion on this?

JohnnyMorganz commented:
Oh yeah... that is a valid point. Ok, then let's leave it with the postponing mechanism. Not sure if there is a nicer way to do it.

Although at this point, is it worth the `ensureConfigured` call inside of every method - maybe we should do the postponing directly in the on request loop? And postpone everything until configured.

My only fear is what if the client never responds with config, and we are stuck forever. But tbh, maybe that just means the client is buggy

voidedWarranties commented on CHANGELOG.md:
```diff
@@ -8,12 +8,18 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ### Deprecated
```
Perhaps only need to document the `--platform` flag as Roblox remains the default?

JohnnyMorganz replied:
Oh I must've been mistaken, I thought "standard" become the default (maybe it should? But that's a breaking change - let's leave it for now)

JohnnyMorganz reviewed (commented):
Sorry to keep delaying this. I think I am happy with the general setup, and we can always refactor it later.

My only concern still lies with the postponing mechanism. I think there is a potential issue here with the ordering of events. Consider the following sequence:
- client requests document diagnostics. Server puts it into postponing queue as not received config
- client sends `textDocument/didUpdate` notification to change document text
- client sends configuration
- server responds to the original diagnostics message. However, the text document is not in the state when the message was sent (it got updated). Now the diagnostics provided are not in sync with the request sent.

I think we might be able to get around this just by also postponing notifications as well as requests (making sure we still allow any relevant notifications?). It is somewhat disappointing that the server completely halts whilst we wait for the config though.

> I did consider this; I believe the reason I didn't is that it seems impossible to use this method for more than one workspace with different platforms. Do you have any suggestion on this?

Also I was rethinking this. Can't we just send the config for each workspace in initialisation options?

Maybe we can structure it as follows:

```js
initializationOptions: {
    "config": {
        "global": /* non-workspace specific config. We can use this if client doesn't support workspace folders */
        [workspaceUri]: /* workspace config */
    }
}
```

Other than this though, I'm happy to proceed. I will try to find some time to test everything that I can think of works on my end (the cons of manual testing...)

voidedWarranties commented:
> I think there is a potential issue here with the ordering of events.

I am not so familiar with the LSP spec, but does [this](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#implementationConsiderations) mean that the client is responsible for handling this scenario and not the server?

> Can't we just send the config for each workspace in initialisation options?

Isn't it possible to add another workspace after initialization?

JohnnyMorganz commented:
The potentially problematic part comes from the second part:

> keep the request running if the client can still make use of the result by, for example, transforming it to a new result by applying the state change to the result.
servers should therefore not decide by themselves to cancel requests simply due to that fact that a state change notification is detected in the queue. As said the result could still be useful for the client.

Although I do not know of any requests that do this in practice.

> Isn't it possible to add another workspace after initialization?

Yeah..., that is also valid. Maybe this concern is for nothing. Let's leave the postponing mechanism for now and I can revisit it if necessary

JohnnyMorganz commented:
We should look into merging this ASAP to prevent it from diverging to much. Are you happy with everything (can it exit draft mode?)

I still need to do my own local testing, especially the points you mentioned in OP

voidedWarranties commented:
The thing blocking moving out of draft is making appropriate changes to the VSCode extension. If that's best done in another PR then I think the server side is basically ready to exit draft.

voidedWarranties commented:
I've finally addressed the extension and split it in a way I think makes sense. The second commit I added updates config and config usage; I think this needs some attention because the resolution between deprecated and new config options is somewhat dissatisfying (sometimes, the old options are ignored entirely and I don't see a better way to resolve it).

Again, things are mostly lift-and-shift, but again, I haven't tested functionality in-depth.

This PR is pretty much feature-complete, so I marked as ready for review.

JohnnyMorganz commented:
Hey, thank you!

Very sorry for the delay. I've been meaning to dedicate some time to finish reviewing this but haven't got round to it yet.

I will try to pick it up next weekend.

JohnnyMorganz commented on editors/code/src/extension.ts:
```diff
@@ -632,7 +242,10 @@ export async function activate(context: vscode.ExtensionContext) {
-      } else if (e.affectsConfiguration("luau-lsp.types")) {
+      } else if (
+        e.affectsConfiguration("luau-lsp.types") ||
+        e.affectsConfiguration("luau-lsp.platform")
```
I think platform here might be too broad - we don't need to reload the whole server if a user decides to enable / disable the Studio plugin, or configure the way the sourcemap works.

JohnnyMorganz reviewed (approved):
I am incredibly sorry in the delay for reviewing this code. Unfortunately, my time to work on this project has become severely limited recently.

I still have not yet had a chance to test this thoroughly locally, as I do not have access to my dev machine for a couple more weeks. However, I really do want to get this merged ASAP so it doesn't become stale any longer.

I am going to cut a new release of the server in its current state, then merge this code in. Then, we can fix-forward any issues we find.

Thank you again for taking the time for doing this, and sorry to make it feel like it was wasted!

JohnnyMorganz commented:
>  I think this needs some attention because the resolution between deprecated and new config options is somewhat dissatisfying (sometimes, the old options are ignored entirely and I don't see a better way to resolve it).

I am wondering two things here:

1) Is it worth moving all the configs under `platform`? It is definitely nice to have, but seems purely cosmetic. I do have more configs I want to move around too, so maybe it could be deferred to a breaking change release.

2) If we do switch over the configs, one thing that I saw rust-analyzer do is to perform a one-time transformation of any deprecated configs to the new ones. Maybe we could do something similar here, to minimise bad config settings. (Of course, won't help non-VSCode users)

voidedWarranties commented:
> I am incredibly sorry in the delay for reviewing this code.

No worries, I think we both have lives outside of this.

> I am wondering two things here:

I think a one-time migration would be nice, and since there isn't a great resolution for configuration right now I've gone ahead and reverted the configuration changes. Those two can be left to another PR.

There are a lot of additional code movements that could happen (e.g. `Utils` and many of the tests), but in the interest of not growing this PR any more I've skipped them.

If you are still interested in reviewing the changes (and I understand if not) it might help (a bit) to color code moved code: `git diff --color-moved upstream/main..HEAD`.

JohnnyMorganz commented on src/Workspace.cpp:
```diff
@@ -289,17 +311,20 @@ void WorkspaceFolder::initialize()
-        if (!isNullWorkspace() && !updateSourceMap())
-        {
-            client->sendWindowMessage(
-                lsp::MessageType::Error, "Failed to load sourcemap.json for workspace '" + name + "'. Instance information will not be available");
-        }
+        isConfigured = true;
```
What happens in this case if the `platform.type` was changed midway through an active session. This check looks like we only mutate registered definitions on startup when we are first not configured, so this wouldn't work I think?

JohnnyMorganz replied:
Actually nvm, I just remembered from the extension code that we need to reload the language server whenever type definitions changes. So this is fine

JohnnyMorganz reviewed (commented):
It's been a while, but I've finally had a chance to sit down and spend some time re-reading this PR. Even though there are a lot of files changed, it actually doesn't seem that bad. The only actual behaviour change I see is the postponing of messages until we are configured.

I've also tested it locally and it seems to be all good. Studio plugin still works as well. I don't have the chance to test in my day-to-day right now, but I think it's fine to move forward.

Just a few final queries (if you remember the context! It's been a long time, so totally fine if not). Then I think we are good to merge.

voidedWarranties commented on src/LanguageServer.cpp:
```diff
@@ -355,35 +406,12 @@ void LanguageServer::processInputLoop()
             {
                 client->sendError(id, JsonRpcException(lsp::ErrorCode::ParseError, e.what()));
             }
-            catch (const std::exception& e)
```
I believe those exception handlers were moved to `handleMessage` and there it should still catch all relevant exceptions.

JohnnyMorganz replied:
Ah yes, makes sense, thanks!

voidedWarranties commented on src/Workspace.cpp:
```diff
@@ -13,9 +17,12 @@ void WorkspaceFolder::openTextDocument(const lsp::DocumentUri& uri, const lsp::D
-    // Mark the file as dirty as we don't know what changes were made to it
-    auto moduleName = fileResolver.getModuleName(uri);
-    frontend.markDirty(moduleName);
+    if (isConfigured)
```
I believe it was done because `getModuleName` now might involve accessing `platform`.

JohnnyMorganz replied:
Cool. I wonder if this can cause any problems where we didn't mark a file dirty appropriately on startup, but I think it is fine for now. I want to get rid of the reliance on platform-specific file paths across the server except for file resolution, so hopefully this won't be a problem in future. Will note it down to resolve: #614

JohnnyMorganz reviewed (approved):
I've cut a release, so we are ready to go. It's been a long time, but your work is much appreciated. Thank you!
